autotx/agents/SwapTokensAgent.py

```diff
@@ -98,58 +83,76 @@ def get_tokens_address(token_in: str, token_out: str, network_info: NetworkInfo)
+
+    return swap_transactions
+
+class BulkSwapTool(AutoTxTool):
+    name: str = "prepare_bulk_swap_transactions"
     description: str = dedent(
         """
-        Prepares a swap transaction for given amount in decimals for given token. Swap should only include the amount for one of the tokens, the other token amount will be calculated automatically.
+        Prepares a batch of buy transactions for given amounts in decimals for given tokens.
         """
     )
 
-    def build_tool(self, autotx: AutoTx) -> Callable[[str, str], str]:
+    def build_tool(self, autotx: AutoTx) -> Callable[[str], str]:
         def run(
-            token_to_sell: Annotated[str, "Token to sell. E.g. '10 USDC' or just 'USDC'"],
-            token_to_buy: Annotated[str, "Token to buy. E.g. '10 USDC' or just 'USDC'"],
+            tokens: Annotated[
+                str, 
+                """
+                Tokens and amounts to swap. ONLY one amount should be provided next to the token symbol PER line.
+                E.g:
+                10 USDC to ETH
+                UNI to 3.3 WBTC
+                1.5 WBTC to USDC
+                """
+            ],
         ) -> str:
-            sell_parts = token_to_sell.split(" ")
-            buy_parts = token_to_buy.split(" ")
-
-            if len(sell_parts) == 2 and len(buy_parts) == 2:
-                return "Invalid input. Only one token amount should be provided. IMPORTANT: Take another look at the user's goal, and try again."
-            
-            if len(sell_parts) < 2 and len(buy_parts) < 2:
-                return "Invalid input. Token amount is missing.\n"
+            swaps = tokens.split("\n")
+            all_txs = []
 
-            token_symbol_to_sell = sell_parts[1] if len(sell_parts) == 2 else sell_parts[0]
-            token_symbol_to_buy = buy_parts[1] if len(buy_parts) == 2 else buy_parts[0]
-
-            exact_amount = sell_parts[0] if len(sell_parts) == 2 else buy_parts[0]
-            amount_symbol = token_symbol_to_sell if len(sell_parts) == 2 else token_symbol_to_buy
-
-            token_in = token_symbol_to_sell.lower()
-            token_out = token_symbol_to_buy.lower()
-            is_exact_input = True if amount_symbol == token_symbol_to_sell else False
-
-            (token_in_address, token_out_address) = get_tokens_address(
-                token_in, token_out, autotx.network
-            )
-
-            swap_transactions = build_swap_transaction(
-                autotx.manager.client,
-                Decimal(exact_amount),
-                token_in_address,
-                token_out_address,
-                autotx.manager.address.hex,
-                is_exact_input,
-            )
-            autotx.transactions.extend(swap_transactions)
+            for swap_str in swaps:
+                (token_to_sell, token_to_buy) = swap_str.strip().split(" to ")
+                txs = swap(autotx, token_to_sell, token_to_buy)
```

Comment:
now that we're looping in the swaps, i think we should gracefully return which transactions has been correctly build and which ones failed (with reason) - when testing this the agent created duplicated transactions because 1 of 3 failed

Reply:
Done